### PR TITLE
support for custom text property for a subtitle

### PR DESCRIPTION
**Pull request by ekuleshov**
Support for custom RichText subtitles https://github.com/brianegan/chewie/issues/479

### CODE REVIEW

File: lib/src/chewie_player.dart [hunk 1]
```diff
@@ -277,7 +277,7 @@ class ChewieController extends ChangeNotifier {
   final List<OptionItem> Function(BuildContext context)? additionalOptions;
 
   /// Define here your own Widget on how your n'th subtitle will look like
-  final Widget Function(BuildContext context, String subtitle)? subtitleBuilder;
+  final Widget Function(BuildContext context, dynamic subtitle)? subtitleBuilder;
```
Comment (Ahmadre) [inline lib/src/chewie_player.dart:280]: `dynamic` is dangerous and should always be avoided (see Flutter Code-Guidelines). If we want to provide a custom `text` and it belongs to the UI, it should be a Widget here, so the User can decide what kind of Widget she/he wants to add here.

File: lib/src/cupertino/cupertino_controls.dart [hunk 2]
```diff
@@ -170,13 +170,18 @@ class _CupertinoControlsState extends State<CupertinoControls>
...
             color: const Color(0x96000000),
             borderRadius: BorderRadius.circular(10.0)),
-        child: Text(
-          currentSubtitle.first!.text,
-          style: const TextStyle(
-            fontSize: 18,
-          ),
-          textAlign: TextAlign.center,
-        ),
+        child: currentSubtitle.first!.text is InlineSpan
```
Comment (Ahmadre) [inline lib/src/cupertino/cupertino_controls.dart:173]: `is` and `as` should be avoided. Instead provide directly the text as a Widget and you don't need this.

File: lib/src/material/material_controls.dart [hunk 3]
```diff
@@ -244,13 +244,18 @@ class _MaterialControlsState extends State<MaterialControls>
...
           borderRadius: BorderRadius.circular(10.0),
         ),
-        child: Text(
-          currentSubtitle.first!.text,
-          style: const TextStyle(
-            fontSize: 18,
-          ),
-          textAlign: TextAlign.center,
-        ),
+        child: currentSubtitle.first!.text is InlineSpan
```
Comment (Ahmadre) [inline lib/src/material/material_controls.dart:247]: see comment in: lib/src/cupertino/cupertino_controls.dart

File: lib/src/material/material_desktop_controls.dart [hunk 4]
```diff
@@ -230,13 +230,18 @@ class _MaterialDesktopControlsState extends State<MaterialDesktopControls>
...
           borderRadius: BorderRadius.circular(10.0),
         ),
-        child: Text(
-          currentSubtitle.first!.text,
-          style: const TextStyle(
-            fontSize: 18,
-          ),
-          textAlign: TextAlign.center,
-        ),
+        child: currentSubtitle.first!.text is InlineSpan
```
Comment (Ahmadre) [inline lib/src/material/material_desktop_controls.dart:233]: see comment in: see comment in: lib/src/cupertino/cupertino_controls.dart

File: lib/src/models/subtitle_model.dart [hunk 5]
```diff
@@ -42,7 +45,10 @@ class Subtitle {
   final int index;
   final Duration start;
   final Duration end;
-  final String text;
+
+  /// Subtitle text. A [String] or another type, e.g. [InlineSpan], [TextSpan], etc.
+  /// Custom types need to be handled by [ChewieController.subtitleBuilder].
+  final dynamic text;
```
Comment (Ahmadre) [inline lib/src/models/subtitle_model.dart:51]: see comment in: lib/src/chewie_player.dart

**Comment by Ahmadre** [comment]
Thank you @ekuleshov for your PR!

File: lib/src/chewie_player.dart (hunk 1 above)
Comment (ekuleshov) [reply lib/src/chewie_player.dart:280]: @Ahmadre using a `Widget` inside of `Subtitle` data object is probably not a good idea, i.e. to keep UI rendering separate from the values.

How about using generic value in `Subtitle<T>` and `Subtitles<T>` classes?

Then `T` could be `String`, `InlineSpan`. That may require to move the `subtitleBuilder` function into `Subtitles` to tie it in to the same generic type. Also this approach won't allow a mix of different text types like in my original PR.

**Comment by ekuleshov** [comment]
> Thank you @ekuleshov for your PR!

You may want to reopen https://github.com/brianegan/chewie/issues/479

**Comment by ekuleshov** [comment]
@Ahmadre I tried to change `Subtitles` and `subtitleBuilder` to be parametrized types like:

```
class Subtitles<T> {
  Subtitles(this.subtitle, { this.subtitleBuilder });

  final List<Subtitle<T>?> subtitle;
  Widget Function(BuildContext context, T subtitle)? subtitleBuilder;
  ...
```

The `subtitle` and `subtitleBuilder` values need to be tied together because of a common generic type, in under `Subtitles` class (which is basically a Typle) stored in the `ChewieController` instance.

Unfortunately I could not make generic types work. Getting a type case issue at runtime even without `is` and `as`.

Now I'm not sure if it is the best way to approach this.

Or perhaps maybe the subtitleBuilder could be changed to take in a `Subtitle` instance instead of a `String`. Then I could add an extension to `Subtitle` class and will be able to do my own type casting in the subtitleBulder implementation.

What do you think?

**Comment by Ahmadre** [comment]
> @Ahmadre I tried to change `Subtitles` and `subtitleBuilder` to be parametrized types like:
> 
> 
> 
> ```
> 
> class Subtitles<T> {
> 
>   Subtitles(this.subtitle, { this.subtitleBuilder });
> 
> 
> 
>   final List<Subtitle<T>?> subtitle;
> 
>   Widget Function(BuildContext context, T subtitle)? subtitleBuilder;
> 
>   ...
> 
> ```
> 
> 
> 
> The `subtitle` and `subtitleBuilder` values need to be tied together because of a common generic type, in under `Subtitles` class (which is basically a Typle) stored in the `ChewieController` instance.
> 
> 
> 
> Unfortunately I could not make generic types work. Getting a type case issue at runtime even without `is` and `as`.
> 
> 
> 
> Now I'm not sure if it is the best way to approach this.
> 
> 
> 
> Or perhaps maybe the subtitleBuilder could be changed to take in a `Subtitle` instance instead of a `String`. Then I could extend `Subtitle` class with my own fields and will be able to do my own type casting in the subtitleBulder implementation.
> 
> 
> 
> What do you think?

thank you for your suggestions. I will have a look at it the next week. I'm on vacation :D

**Comment by Ahmadre** [comment]
LGTM @ekuleshov :) Could you please update your branch with the latest master? After that I would merge this! I really like it and I think this way we got much more flexibility! 

And could you check the CI-build errors and correct them in your branch?

**Comment by ekuleshov** [comment]
> LGTM

Which part? Like I said, I could not make generified variant work.

A variant with extending `Subtitle` turned out rather ugly. Personally I'd have preferred a variant with `dynamic` text.

**Comment by ekuleshov** [comment]
It seems like to make subtitles to use generic value instead of String will require generic type all the way up to the `Chewie` widget.

All in all, it seems like coupling subtitles into every video control implementation is too restrictive. It would be better to decouple subtitles from that and use a separate widget overlay instead. Somewhat similar to what `subtitle_wrapper_package` does. Then generic type only need to be specified when configuring subtitles overlay.

**Comment by Ahmadre** [comment]
I meant the actual changed with dynamic looks good!

We can do that like in the first step.

**Comment by ekuleshov** [comment]
> We can do that like in the first step.

@Ahmadre okay. I re-did the variant with `dynamic` type. Please review

**Comment by ekuleshov** [comment]
@Ahmadre do you have a timeline on getting this change into a plugin release?

**Pull request merged**